### ISSUE TITLE
[Question] batch independent multioutput GP performance with large #outputs

### ISSUE THREAD

**Issue by livermoreee** [cornellius-gp/gpytorch#1920]
Hi, I am currently trying to use the batch independent multioutput GP to model a large number of curves (> 1000). However, the performance doesn't seem to be well for this many curves. I replicate the issue below with an example. 

Below I define the training data. This follows the tutorial of modeling a sine curve, except that train_y is tiled so I am simulating a number of tasks, which in this case is 10; of course in this case each curve is exactly the same since I tiled it. 
```
tasks = 10
train_x = torch.linspace(0, 1, 100)
train_y = torch.sin(train_x * (2 * math.pi)) + torch.randn(train_x.size()) * 0.2
train_y = torch.unsqueeze(train_y, -1)
train_y = torch.tile(train_y, (1, tasks))
```
I then train the batch independent GP with the standard 50 iterations with LR = 0.1, and plot the fitted image for the first dimension, which in this case doesn't make a difference because each curve is the same. 
![Screenshot 2022-02-19 13:48:45](https://user-images.githubusercontent.com/100048907/154820259-2cde965f-730c-443d-8937-60ca310aa74b.png)
Here we can see the performance is good. However, if I increase the number of tasks to 500, here is the result after the same training procedure: 
![Screenshot 2022-02-19 13:51:00](https://user-images.githubusercontent.com/100048907/154820303-fac6e217-bb94-43ae-946c-e4552a41ecbb.png)

And as I further increase it, the result of training becomes worse and worse. When number of tasks is greater than 1000, increasing the training iterations no longer work anymore and loss is stuck at suboptimal values. 

I am wondering why this is the case? Have I misunderstood how batched GP treats each task? Based on the description for the batch independent GP in the docs, I assume that it simply makes use of the block design to train the hyperparameters for each single curve independently, which means the gradients should be independent? However here, it seems like adding a task interferes with other tasks by making learning harder. 

Upon some more digging, I suspect that it might be because of the likelihood, which is a multitask gaussian likelihood that outputs a single scalar, which would lead to the final loss being a scalar and gradients interfering with each other. I am wondering if there is a way to bypass this? Is there a way to output a vector loss (a scalar loss for each curve) to make sure that each GP is actually trained independently? 

Thanks!

**Comment by wjmaddox** [comment]
I'm not entirely sure what the issue is in that tutorial but this seems to be a work-around (and is probably the most commonly used way of using batched GPs right now):

```python
class BatchIndependentMultitaskGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super().__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean(batch_shape=torch.Size([tasks]))
        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.RBFKernel(batch_shape=torch.Size([tasks])),
            batch_shape=torch.Size([tasks])
        )

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
#         return gpytorch.distributions.MultitaskMultivariateNormal.from_batch_mvn(
#             gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
#         )
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

train_y = train_y.t() # want batch x n rather than n x batch

# likelihood = gpytorch.likelihoods.MultitaskGaussianLikelihood(num_tasks=tasks, has_global_noise=False)
likelihood = gpytorch.likelihoods.GaussianLikelihood(batch_shape=torch.Size([tasks]))
model = BatchIndependentMultitaskGPModel(train_x, train_y, likelihood)
```

Then you train by summing across the batch dimension.

**Comment by livermoreee** [comment]
Thank you! That indeed solves the issue. Closing it as resolved.
